File: VB/XpoClasses.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo
Imports System.ComponentModel
Imports System.Collections.Generic

Namespace ManyToMany
	<NonPersistent> _
	Public Class MyPersistentBase
		Inherits PersistentBase
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		<Key(True)> _
		Public Oid As Integer
	End Class

	Public Class User
		Inherits MyPersistentBase
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _FullName As String
		Public Property FullName() As String
			Get
				Return _FullName
			End Get
			Set(ByVal value As String)
				SetPropertyValue("FullName", _FullName, value)
			End Set
		End Property

		<Association, Browsable(False)> _
		Public ReadOnly Property UserToGroupLinks() As IList(Of UserToGroupLink)
			Get
				Return GetList(Of UserToGroupLink)("UserToGroupLinks")
			End Get
		End Property

		<ManyToManyAlias("UserToGroupLinks", "LinkGroup")> _
		Public ReadOnly Property MemberOfGroups() As IList(Of Group)
			Get
				Return GetList(Of Group)("MemberOfGroups")
			End Get
		End Property
	End Class
	Public Class Group
		Inherits MyPersistentBase
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _Name As String
		Public Property Name() As String
			Get
				Return _Name
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Name", _Name, value)
			End Set
		End Property

		<Association, Browsable(False)> _
		Public ReadOnly Property UserToGroupLinks() As IList(Of UserToGroupLink)
			Get
				Return GetList(Of UserToGroupLink)("UserToGroupLinks")
			End Get
		End Property

		<ManyToManyAlias("UserToGroupLinks", "LinkUser")> _
		Public ReadOnly Property Members() As IList(Of User)
			Get
				Return GetList(Of User)("Members")
			End Get
		End Property
	End Class

	Public Class UserToGroupLink
		Inherits MyPersistentBase
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private _LinkUser As User
		<Association> _
		Public Property LinkUser() As User
			Get
				Return _LinkUser
			End Get
			Set(ByVal value As User)
				SetPropertyValue("LinkUser", _LinkUser, value)
			End Set
		End Property
		Private _LinkGroup As Group
		<Association> _
		Public Property LinkGroup() As Group
			Get
				Return _LinkGroup
			End Get
			Set(ByVal value As Group)
				SetPropertyValue("LinkGroup", _LinkGroup, value)
			End Set
		End Property
	End Class

End Namespace

File: CS/ManyToMany/XpoClasses.cs
using System;
using DevExpress.Xpo;
using System.ComponentModel;
using System.Collections.Generic;

namespace ManyToMany {
    [NonPersistent]
    public class MyPersistentBase : PersistentBase {
        public MyPersistentBase(Session session) : base(session) { }

        [Key(true)]
        public int Oid;
    }

    public class User : MyPersistentBase {
        public User(Session session) : base(session) { }

        private string _FullName;
        public string FullName {
            get {
                return _FullName;
            }
            set {
                SetPropertyValue("FullName", ref _FullName, value);
            }
        }

        [Association, Browsable(false)]
        public IList<UserToGroupLink> UserToGroupLinks {
            get {
                return GetList<UserToGroupLink>("UserToGroupLinks");
            }
        }

        [ManyToManyAlias("UserToGroupLinks", "LinkGroup")]
        public IList<Group> MemberOfGroups {
            get {
                return GetList<Group>("MemberOfGroups");
            }
        }
    }
    public class Group : MyPersistentBase {
        public Group(Session session) : base(session) { }

        private string _Name;
        public string Name {
            get {
                return _Name;
            }
            set {
                SetPropertyValue("Name", ref _Name, value);
            }
        }

        [Association, Browsable(false)]
        public IList<UserToGroupLink> UserToGroupLinks {
            get {
                return GetList<UserToGroupLink>("UserToGroupLinks");
            }
        }

        [ManyToManyAlias("UserToGroupLinks", "LinkUser")]
        public IList<User> Members {
            get {
                return GetList<User>("Members");
            }
        }
    }

    public class UserToGroupLink : MyPersistentBase {
        public UserToGroupLink(Session session) : base(session) { }

        private User _LinkUser;
        [Association]
        public User LinkUser {
            get { return _LinkUser; }
            set { SetPropertyValue("LinkUser", ref _LinkUser, value); }
        }
        private Group _LinkGroup;
        [Association]
        public Group LinkGroup {
            get { return _LinkGroup; }
            set { SetPropertyValue("LinkGroup", ref _LinkGroup, value); }
        }
    }

}
